### PR TITLE
Tpm for aarch64

### PR DESCRIPTION
**Pull request by ashwing123**
Added a test for qemu-system-aarch64 full string with a tpm device included. Full string has been tested on arm64 server, after creating a socket via:
`swtpm socket --tpmstate=dir=./ --ctr=type=unixio,path=tpm.sock --log=level=20,file="log" --pid=file=“pid” --tpm2`

### CODE REVIEW

File: device_aarch64.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// +build aarch64 arm64
+
+package qcli
+
+const TPMTISDeviceAarch64 DeviceDriver = "tpm-tis-device"
```
Comment (raharper) [inline device_aarch64.go]: How unfortunate, x86 has a tpm-tis device as well, but called 'tpm-tis';

Comment (raharper) [inline device_aarch64.go]:
```suggestion
const TPMTISDevice DeviceDriver = "tpm-tis-device"
```

Since this is only compiled in for aarch64, no need to include the arch name in the string.

**Review by raharper — CHANGES_REQUESTED** [review]
This PR will reduce in since one the main aarch PR lands and you rebase this branch on the new tip.

File: device_aarch64.go (hunk 1 above)
Comment (ashwing123) [reply device_aarch64.go]: Would this conflict with the `TPMTISDevice` in the regular `device.go` file (if both are built)

Comment (raharper) [reply device_aarch64.go]: > Would this conflict with the `TPMTISDevice` in the regular `device.go` file (if both are built)

Correct, but since you're marking this file with // build +aarch64 ; it will only get included in the arm build.

Comment (raharper) [reply device_aarch64.go]: > > Would this conflict with the `TPMTISDevice` in the regular `device.go` file (if both are built)
>
> Correct, but since you're marking this file with // build +aarch64 ; it will only get included in the arm build.

Ah, then I guess we'd also need to push the TPMTisDevice into a device_amd64.go.

Here, I think let's leave this TPMTISDeviceAarch64 out; and instead tpm.go, we can use a switch and goarch

```golang
func (tpm TPMDevice) deviceName() string {
    switch tpm.Driver {
    case TPMTISDevice:
          if runtime.GOARCH == "aarch64" || runtime.GOARCH == "arm64" {
              return string(tpm.Driver + "-device")
          }
    }
    return string(tpm.Driver)
}
```

Comment (ashwing123) [reply device_aarch64.go]: Yep this makes sense, will include. This may mess up some of the arch dependent full string tests though (ex. in qemu_test.go for the tpm_socket so I'll try make sure everything's compatible

File: qemu_test.go [hunk 2]
```diff
@@ -871,6 +872,10 @@ func TestFullUEFISpiceMachineCommand(t *testing.T) {
 }
 
 func TestFullUEFITPMMachineCommand(t *testing.T) {
+	//don't need to run test if we have aarch64 -- corresponding test can be found in qemu_aarch64_test.go
+	if runtime.GOARCH == "aarch64" || runtime.GOARCH == "arm64" {
+		return
+	}
```
Comment (raharper) [inline qemu_test.go]: I was thinking we could use `runtime.GOARCH` to select both the config generated and the expected cmdline?

```
var c *Config
u := UEFIFirmareDevice{Code: "", Vars: "uefi-nvram.fd"}

switch runtime.GOARCH {
case "aarch64", "arm64":
     c = fullVMConfigAarch()
     u.Code = "/usr/share/AAVMF/AAMF_CODE.fd"
     expected = fullUefiVMTPMAarch64
case "x86_64":
     c = fullVMConfig()
     u.Code = "/usr/share/OVMF/OVMF_CODE.fd"
     expected = fullUefiVMTPM
}
```

The tpm code should be the same since the Driver value will be set correctly
with the runtime switch there.

**Review by raharper — CHANGES_REQUESTED** [review]

File: qemu_test.go (hunk 2 above)
Comment (ashwing123) [reply qemu_test.go]: This makes a lot more sense. I was actually thinking I could try call the other test function somehow but this should work better

**Comment by ashwing123** [comment]
Just moved machine type "virt" into `machine.go` and got rid of aarch specific machine so hopefully the build will work now

File: blockdev_test.go [hunk 3]
```diff
@@ -6,7 +6,7 @@ var (
 	deviceBlockString         = "-drive file=/var/lib/vm.img,id=hd0,if=none,format=qcow2,aio=threads,cache=unsafe,discard=unmap,detect-zeroes=unmap,readonly=on -device virtio-blk-pci,drive=hd0,serial=abc-123,disable-modern=true,addr=0x03,bus=pcie.0,logical_block_size=4096,physical_block_size=4096,scsi=off,config-wce=off,romfile=efi-virtio.rom,share-rw=on"
 	deviceBlockAddrString     = "-drive file=/var/lib/vm.img,id=hd0,if=none,format=qcow2 -device virtio-blk-pci,drive=hd0,serial=hd0,disable-modern=false,addr=0x07,bus=pcie.0,scsi=off,config-wce=off"
 	deviceBlockPFlashROString = "-drive file=/usr/share/OVMF/OVMF_CODE.fd,id=pflash0,if=pflash,format=raw,readonly=on"
-	deviceBlockPFlashRWString = "-drive file=uefi_nvram.fd,id=pflash1,if=pflash,format=raw"
+	deviceBlockPFlashRWString = "-drive file=uefi-nvram.fd,id=pflash1,if=pflash,format=raw"
```
Comment (raharper) [inline blockdev_test.go]: Let's leave these as they are; it's an unrelated change to enabling tpm on arm support

File: blockdev_test.go [hunk 4]
```diff
@@ -134,7 +134,7 @@ func TestAppendDeviceBlockPFlashRW(t *testing.T) {
 	blkdev := BlockDevice{
 		Driver:    PFlash,
 		ID:        "pflash1",
-		File:      "uefi_nvram.fd",
+		File:      "uefi-nvram.fd",
```
Comment (raharper) [inline blockdev_test.go]: Let's leave these as they are; it's an unrelated change to enabling tpm on arm support

File: example/run-vms.go [hunk 5]
```diff
@@ -69,7 +69,7 @@ func getFullConfig() qcli.Config {
 			qcli.BlockDevice{
 				Driver:    qcli.PFlash,
 				ID:        "pflash1",
-				File:      "uefi_nvram.fd",
+				File:      "uefi-nvram.fd",
```
Comment (raharper) [inline example/run-vms.go]: Let's leave these as they are; it's an unrelated change to enabling tpm on arm support

File: machine.go [hunk 6]
```diff
@@ -85,6 +85,7 @@ const (
 	MachineTypeMicrovm string = "microvm"
 	MachineTypePC35    string = "q35"
 	MachineTypePC      string = "pc"
+	MachineTypeVirtAarch64 string = "virt"
```
Comment (raharper) [inline machine.go]:
```suggestion
	MachineTypeVirt string = "virt"
```

I think it's fine to just call this MachineTypeVirt -- we can runtime check MachineType compat with host arch.

File: machine_aarch64_test.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// +build aarch64 arm64
```
Comment (raharper) [inline machine_aarch64_test.go]: Shouldn't we drop this and move this test into machine_test.go ?  There's nothing really arch specific here, only that at runtime qemu-system-$arch may or maynot support this machine type.

File: qemu_test.go [hunk 8]
```diff
@@ -699,10 +700,12 @@ func TestBadCPUs(t *testing.T) {
...
+	fullUefiTPMAarch64VM = "-machine virt,accel=kvm -cpu host -m 1G -chardev socket,id=chrtpm0,path=tpm.sock -tpmdev emulator,id=tpm0,chardev=chrtpm0 -device tpm-tis-device,tpmdev=tpm0 -drive file=udisk.img,id=hd0,if=none,format=qcow2 -device virtio-blk-pci,drive=hd0,serial=hd0,disable-modern=false,addr=0x1e,bus=pcie.0,scsi=off,config-wce=off -drive file=ubuntu-22.04.2-live-server-arm64.iso,id=cdrom0,if=none,format=raw,media=cdrom,readonly=on -device virtio-blk-pci,drive=cdrom0,serial=cdrom0,bootindex=0,disable-modern=false,addr=0x1d,bus=pcie.0,scsi=off,config-wce=off -drive if=pflash,format=raw,readonly=on,file=/usr/share/AAVMF/AAVMF_CODE.ms.fd -drive if=pflash,format=raw,file=uefi-nvram.fd -object memory-backend-ram,id=dimm1,size=1G -numa node,memdev=dimm1 -nographic"
```
Comment (raharper) [inline qemu_test.go]:
```suggestion
	fullUefiAarch64VMTPM = "-machine virt,accel=kvm -cpu host -m 1G -chardev socket,id=chrtpm0,path=tpm.sock -tpmdev emulator,id=tpm0,chardev=chrtpm0 -device tpm-tis-device,tpmdev=tpm0 -drive file=udisk.img,id=hd0,if=none,format=qcow2 -device virtio-blk-pci,drive=hd0,serial=hd0,disable-modern=false,addr=0x1e,bus=pcie.0,scsi=off,config-wce=off -drive file=ubuntu-22.04.2-live-server-arm64.iso,id=cdrom0,if=none,format=raw,media=cdrom,readonly=on -device virtio-blk-pci,drive=cdrom0,serial=cdrom0,bootindex=0,disable-modern=false,addr=0x1d,bus=pcie.0,scsi=off,config-wce=off -drive if=pflash,format=raw,readonly=on,file=/usr/share/AAVMF/AAVMF_CODE.ms.fd -drive if=pflash,format=raw,file=uefi-nvram.fd -object memory-backend-ram,id=dimm1,size=1G -numa node,memdev=dimm1 -nographic"
```

Comment (raharper) [reply qemu_test.go]: The pattern here is full<firmware><arch>VM<variant>

File: qemu_test.go [hunk 9]
```diff
@@ -828,16 +831,60 @@ func fullVMConfig() *Config {
...
+	return c
+}
 
-	u := UEFIFirmwareDevice{
-		Code: "/usr/share/OVMF/OVMF_CODE.fd",
-		Vars: "uefi_nvram.fd",
+
+func TestFullUEFIMachineCommand(t *testing.T) {
+	var c *Config
+	u := UEFIFirmwareDevice{Code: "", Vars: "uefi-nvram.fd",}
```
Comment (raharper) [inline qemu_test.go]:
```suggestion
	u := UEFIFirmwareDevice{Code: "", Vars: "uefi_nvram.fd",}
```

File: qemu_test.go [hunk 10]
```diff
@@ -853,7 +900,7 @@ func TestFullUEFISpiceMachineCommand(t *testing.T) {
 
 	u := UEFIFirmwareDevice{
 		Code: "/usr/share/OVMF/OVMF_CODE.fd",
-		Vars: "uefi_nvram.fd",
+		Vars: "uefi-nvram.fd",
```
Comment (raharper) [inline qemu_test.go]:
```suggestion
		Vars: "uefi_nvram.fd",
```

File: qemu_test.go [hunk 11]
```diff
@@ -871,22 +918,29 @@ func TestFullUEFISpiceMachineCommand(t *testing.T) {
...
 
 func TestFullUEFITPMMachineCommand(t *testing.T) {
-	c := fullVMConfig()
-
-	u := UEFIFirmwareDevice{
-		Code: "/usr/share/OVMF/OVMF_CODE.fd",
-		Vars: "uefi_nvram.fd",
-	}
+	var c *Config
+	u := UEFIFirmwareDevice{Code: "", Vars: "uefi-nvram.fd",}
```
Comment (raharper) [inline qemu_test.go]:
```suggestion
	u := UEFIFirmwareDevice{Code: "", Vars: "uefi_nvram.fd",}
```

File: qemu_test.go [hunk 12]
```diff
@@ -871,22 +918,29 @@ func TestFullUEFISpiceMachineCommand(t *testing.T) {
...
+		expected = fullUefiVMTPM
+	}
+	
 	c.UEFIFirmwareDevices = append(c.UEFIFirmwareDevices, u)
 
 	c.TPM = TPMDevice{
 		ID:     "tpm0",
 		Driver: TPMTISDevice,
-		Path:   "tpm.socket",
+		Path:   "tpm.sock",
```
Comment (raharper) [inline qemu_test.go]: Let's skip this sort of change and just update the expected command line to match the existing constructed devices.

File: tpm_test.go [hunk 13]
```diff
@@ -26,11 +27,16 @@ var ValidTPM = TPMDevice{
 }
 
 func TestTPMDevice(t *testing.T) {
+	chardevStr := "-chardev socket,id=chrtpm0,path=tpm.socket -tpmdev emulator,id=tpm0,chardev=chrtpm0 -device tpm-tis,tpmdev=tpm0"
+	if runtime.GOARCH == "aarch64" || runtime.GOARCH == "arm64" {
+		chardevStr = "-chardev socket,id=chrtpm0,path=tpm.socket -tpmdev emulator,id=tpm0,chardev=chrtpm0 -device tpm-tis-device,tpmdev=tpm0"
+	}
+
```
Comment (raharper) [inline tpm_test.go]:
```suggestion
    chardevName := "tpm-tis"
	if runtime.GOARCH == "aarch64" || runtime.GOARCH == "arm64" {
		chardevName = "tpm-tis-device"
	}
	chardevStr := fmt.Sprintf("-chardev socket,id=chrtpm0,path=tpm.socket -tpmdev emulator,id=tpm0,chardev=chrtpm0 -device %s,tpmdev=tpm0", chardevName)
```

It's easier to see what part of the expected string is different without repeating the longer string in both cases.

**Review by raharper — CHANGES_REQUESTED** [review]

**Review by raharper — APPROVED** [review]
